### PR TITLE
snowflake - fixes and tests

### PR DESCRIPTION
**Pull request by mlorek**
remaining TOKENs as IDs will be added in a separate PR

### CODE REVIEW

File: sql/snowflake/SnowflakeParser.g4 [hunk 1]
```diff
@@ -1639,21 +1661,27 @@ function_definition
...
+    : CREATE or_replace? SECURE? FUNCTION if_not_exists? object_name LR_BRACKET ( arg_decl (COMMA  arg_decl)* )? RR_BRACKET
+        RETURNS (data_type | TABLE LR_BRACKET (col_decl (COMMA col_decl)* )? RR_BRACKET)
+        (LANGUAGE (JAVA | PYTHON | JAVASCRIPT | SQL))?
+        (CALLED ON NULL_ INPUT | RETURNS NULL_ ON NULL_ INPUT | STRICT)?
+        (VOLATILE | IMMUTABLE)?
+        (PACKAGES EQ '(' string_list ')' )?
+        (RUNTIME_VERSION EQ (string | FLOAT))?
+        (IMPORTS EQ '(' string_list ')' )?
+        (PACKAGES EQ '(' string_list ')' )?
+        (HANDLER EQ string )?
```
Comment (MasterKuat) [inline sql/snowflake/SnowflakeParser.g4:1673]: @mlorek  carefull, "Packages", "runtime",handler etc are only valid with python/java/scala

File: sql/snowflake/SnowflakeParser.g4 [hunk 2]
```diff
@@ -1731,22 +1759,42 @@ procedure_definition
...
-    | CREATE or_replace? SECURE? PROCEDURE object_name LR_BRACKET (arg_decl (COMMA arg_decl)*)? RR_BRACKET RETURNS data_type (
-        NOT NULL_
-    )? LANGUAGE JAVASCRIPT (CALLED ON NULL_ INPUT | RETURNS NULL_ ON NULL_ INPUT | STRICT)? (
-        VOLATILE
-        | IMMUTABLE
-    )? // Note: VOLATILE and IMMUTABLE are deprecated.
-    comment_clause? executa_as? AS procedure_definition
+    : CREATE or_replace? PROCEDURE object_name LR_BRACKET (arg_decl (COMMA arg_decl)*)? RR_BRACKET
+        RETURNS (data_type | TABLE LR_BRACKET (col_decl (COMMA col_decl)*)? RR_BRACKET)
+        not_null?
```
Comment (MasterKuat) [inline sql/snowflake/SnowflakeParser.g4:1769]: Even if it's not in doc, NULL is also allowed (at least for SQL)

File: sql/snowflake/SnowflakeParser.g4 [hunk 3]
```diff
@@ -2308,10 +2356,16 @@ create_table
...
+    ;
+
 create_table_clause
-    : '(' column_decl_item_list ')' cluster_by? stage_file_format? (
+    : ( column_decl_item_list_paren cluster_by?
+      | cluster_by? comment_clause? column_decl_item_list_paren
+      ) stage_file_format? (
         STAGE_COPY_OPTIONS EQ LR_BRACKET copy_options RR_BRACKET
-    )? (DATA_RETENTION_TIME_IN_DAYS EQ num)? (MAX_DATA_EXTENSION_TIME_IN_DAYS EQ num)? change_tracking? default_ddl_collation? copy_grants?
+    )? (DATA_RETENTION_TIME_IN_DAYS EQ num)? (MAX_DATA_EXTENSION_TIME_IN_DAYS EQ num)? change_tracking? default_ddl_collation? copy_grants? comment_clause?
```
Comment (MasterKuat) [inline sql/snowflake/SnowflakeParser.g4:2368]: comment_clause is already in create_table

**Comment by teverett** [comment]
@mlorek  there are merge conflicts.

File: sql/snowflake/SnowflakeParser.g4 (hunk 1 above)
Comment (mlorek) [reply sql/snowflake/SnowflakeParser.g4:1672]: functions will be refactored later as there are many discrepancies in the official doc

File: sql/snowflake/SnowflakeParser.g4 (hunk 3 above)
Comment (mlorek) [reply sql/snowflake/SnowflakeParser.g4:2367]: comment clause can be placed in 2 different places, this needs refactoring

**Comment by mlorek** [comment]
> @mlorek there are merge conflicts.
@teverett ready for a review now

File: sql/snowflake/SnowflakeParser.g4 [hunk 4]
```diff
@@ -3427,15 +3482,68 @@ non_reserved_words
...
     | JAVASCRIPT
+    | LAST_NAME
+    | LAST_QUERY_ID
+    | LEAD
+    | LOCAL
+    | MAX_CONCURRENCY_LEVEL
+    | NAME
+    | NULLIF
+    | NVL
+    | NVL2
```
Comment (MasterKuat) [inline sql/snowflake/SnowflakeParser.g4:3514]: Already in (binary_builtin_function/ternary_builtin_function)

File: sql/snowflake/SnowflakeParser.g4 (hunk 3 above)
Comment (MasterKuat) [reply sql/snowflake/SnowflakeParser.g4:2367]: Hum I see, thank you

File: sql/snowflake/SnowflakeParser.g4 [hunk 5]
```diff
@@ -1138,30 +1141,45 @@ clustering_action
...
 
+add_col_def
+    : column_name data_type (NOT? NULL_)?
+              default_value2?
+              inline_constraint?
+              (WITH? MASKING POLICY id_ (USING '(' column_name COMMA column_list ')' )? )?
+              (COMMENT string)?
+    ;
+
+default_value2
```
Comment (MasterKuat) [inline sql/snowflake/SnowflakeParser.g4:1152]: Why not using existent default_value ? 
```ALTER TABLE  TESTSEED ADD ident int IDENTITY START = 2 INCREMENT BY 1;``` is also accepted

File: sql/snowflake/SnowflakeParser.g4 [hunk 6]
```diff
@@ -1138,30 +1141,45 @@ clustering_action
     | DROP CLUSTERING KEY
     ;
 
+add_col_def
+    : column_name data_type (NOT? NULL_)?
```
Comment (MasterKuat) [inline sql/snowflake/SnowflakeParser.g4:1145]: ```null_not_null?``` ?

**Comment by mlorek** [comment]
@kaby76 can the go build be rerun?

**Comment by kaby76** [comment]
> @kaby76 can the go build be rerun?

I'm not an admin. To re-execute the workflow, an admin needs to go to the page for the last Action of the PR, then click on "Re-run all jobs". (I don't know if you have access to that UI of github.) Otherwise, you have to make a change to something, commit, and push the commit for your PR.

Besides, there is no way to execute only the Go target checks. The workflow determines the extent of the change, but retests all targets because I don't think it mattered figuring out a finer level of granularity of what to test (e.g., I change only a base class for the Go target, so, retest that). Tests are parallelized across machines and targets, so I didn't see the need.

**Comment by teverett** [comment]
@mlorek  thanks!

**Comment by KvanTTT** [comment]
> To re-execute the workflow, an admin needs to go to the page for the last Action of the PR, then click on "Re-run all jobs".

It's also possible just to force-push commits to trigger remote build. No admin action is required in this case.

**Pull request merged**